crypto/evp/e_eea3.c

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2018 BaishanCloud. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "internal/cryptlib.h"
+
+#ifndef OPENSSL_NO_ZUC
+
+# include <openssl/evp.h>
+# include <openssl/objects.h>
+
+# include "internal/zuc.h"
+# include "internal/evp_int.h"
+
+typedef struct {
+    ZUC_KEY zk;                 /* working key */
+} EVP_EEA3_KEY;
+
+# define data(ctx) ((EVP_EEA3_KEY *)EVP_CIPHER_CTX_get_cipher_data(ctx))
+
+static int eea3_init_key(EVP_CIPHER_CTX *ctx, const unsigned char *key,
+                        const unsigned char *iv, int enc);
+static int eea3_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
+                      const unsigned char *in, size_t inl);
+static int eea3_cleanup(EVP_CIPHER_CTX *ctx);
+
+static const EVP_CIPHER zuc_128_eea3_cipher = {
+    NID_zuc_128_eea3,
+    1, EVP_ZUC_KEY_SIZE, 5,
+    EVP_CIPH_VARIABLE_LENGTH,
+    eea3_init_key,
+    eea3_cipher,
+    eea3_cleanup,
+    sizeof(EVP_EEA3_KEY),
+    NULL,
+    NULL,
+    NULL,
+    NULL
+};
+
+const EVP_CIPHER *EVP_eea3(void)
+{
+    return &zuc_128_eea3_cipher;
+}
+
+static int eea3_init_key(EVP_CIPHER_CTX *ctx, const unsigned char *key,
+                         const unsigned char *iv, int enc)
+{
+    EVP_EEA3_KEY *ek = data(ctx);
+    ZUC_KEY *zk = &ek->zk;
+    uint32_t count;
+    uint32_t bearer;
+    uint32_t direction;
+
+    zk->k = key;
+
+    /*
+     * This is a lazy approach: we 'borrow' the 'iv' parameter
+     * to use it as a place of transfer the EEA3 iv params -
+     * count, bearer and direction.
+     *
+     * count is 32 bits, bearer is 5 bits and direction is 1
+     * bit so we read the first 38 bits of iv. And the whole
+     * iv is set to 5 bytes (40 bits).
+     */
```

Comment:
This seems very adhoc. Can you point to a standard that defines this "packed" IV format?

Reply:
It *is* ad-hoc...  the first part of the comment suggests that at least.

Another way would be to invent a few controls for ZUC with which one could give the individual parameters.

Reply:
> Can you point to a standard that defines this "packed" IV format?

Hi @bbbrumley , actually there is no such standard yet. The standard of EEA3 defines that the 3 parameters are used to form the final IV. So I think this is a way of passing those parameters by using EVP in OpenSSL (at least those 3 parameters have relations with the final IV).

Another way is as what @levitte said, it's also possible to pass those 3 parameters by other method, as EVP_ctrl I think...

I don't know which one is better, but it seems current method could avoid changes to other part of OpenSSL, say the `openssl enc` command...